src/shared/ui/tag-box.tsx

```diff
@@ -0,0 +1,138 @@
+import {
+  cn,
+  isEmptyString,
+  isPopulatedArray,
+  isPopulatedString,
+  mergeUniqueStrings,
+} from "@/shared/libs";
+import { XCircle } from "lucide-react";
+import React, { forwardRef, useImperativeHandle, useState } from "react";
+import { Badge } from "./badge";
+import { Separator } from "./separator";
+
+export type TagBoxHandle = {
+  set: (newTags: string[]) => void;
+};
+
+type ValidationPattern = {
+  message: string;
+  regex: RegExp;
+};
+
+export interface TagBoxProps
+  extends React.InputHTMLAttributes<HTMLInputElement> {
+  onTagsChange?: (tags: string[]) => void;
+  initialTags?: string[];
+  patterns?: ValidationPattern[];
+}
+
+const TagBox = forwardRef<TagBoxHandle, TagBoxProps>(
+  (
+    {
+      className,
+      initialTags = [],
+      onTagsChange,
+      patterns = [],
+      placeholder = "Type and press Enter",
+    },
+    ref
+  ) => {
+    const [tags, setTags] = useState<string[]>(initialTags);
```

Comment:
1. Maybe make generic value T instead of just string?
2. We can use Set<T> in this useState, not in mergeUniqueStrings. This should contain all logic related to removing duplicate values.